### PR TITLE
Move mi_loss() random sampling functions to deepali.core

### PR DESCRIPTION
**Pull request by aschuh-hf**
Following up on #1 by @qiuhuaqi , moving random sampling functions to `deepali.core` and generalizing these.

### CODE REVIEW

File: deepali/losses/functional.py [hunk 1]
```diff
@@ -737,32 +671,56 @@ def mi_loss(
         sample_ratio: Ratio of voxels in the image domain randomly sampled to compute the loss.
         num_samples: Number of voxels in the image domain randomly sampled to compute the loss,
             ignored if `sample_ratio` is also set.
-        normalized: Calculate Normalized Mutual Information instead of Mutual Information if True.
```
Comment (schuhschuh) [inline deepali/losses/functional.py:740]: Accidentally removed.

File: deepali/core/image.py [hunk 2]
```diff
@@ -1163,6 +1164,75 @@ def grid_sample_mask(
...
+    if not replacement and mask is None:
+        perm = torch.empty(numel, dtype=torch.int64, device=device)
+        index = torch.empty((shape[0], num_samples), dtype=torch.int64, device=device)
+        for row in index:
+            torch.randperm(numel, generator=generator, out=perm, device=device)
+            row.copy_(perm.narrow(0, 0, num_samples), non_blocking=True)
+        out = [x.index_select()[:, :, index] for x in input]
+    else:
+        if mask is None:
+            mask = torch.ones((1, 1, numel), dtype=torch.float32, device=device)
```
Comment (schuhschuh) [inline deepali/core/image.py:1219]: Use `torch.randint` instead of `torch.randperm` in `if`-block instead of going through `multinomial`.

File: deepali/core/image.py [hunk 3]
```diff
@@ -1163,6 +1164,75 @@ def grid_sample_mask(
...
+        raise ValueError("rand_sample() 'data' tensors must have identical spatial shape")
+    numel = shape[2:].numel()
+    if not replacement and num_samples > numel:
+        raise ValueError("rand_sample() 'num_samples' is greater than number of spatial points")
+    input = [x.flatten(2) for x in input]
+    if not replacement and mask is None:
+        perm = torch.empty(numel, dtype=torch.int64, device=device)
+        index = torch.empty((shape[0], num_samples), dtype=torch.int64, device=device)
+        for row in index:
+            torch.randperm(numel, generator=generator, out=perm, device=device)
```
Comment (schuhschuh) [inline deepali/core/image.py:1214]: Considering timings reported at https://discuss.pytorch.org/t/torch-equivalent-of-numpy-random-choice/16146/19, maybe best to perform `randperm` on CPU and/or use one of the faster alternative approaches noted.

File: deepali/core/image.py [hunk 4]
```diff
@@ -1163,6 +1164,75 @@ def grid_sample_mask(
...
+            row.copy_(perm.narrow(0, 0, num_samples), non_blocking=True)
+        out = [x.index_select()[:, :, index] for x in input]
```
Comment (qiuhuaqi) [inline deepali/core/image.py:1216]: this `index_select()[:,:,index]` doesn't look right?

Comment (schuhschuh) [reply deepali/core/image.py:1216]: Good catch. Forgot to delete this line.

File: deepali/core/image.py [hunk 5]
```diff
@@ -1163,6 +1164,81 @@ def grid_sample_mask(
...
+        raise ValueError("rand_sample() 'data' must be one or more tensors of shape (N, C, ..., X)")
+    shape = input[0].shape
+    device = input[0].device
+    if any(x.shape[2:] != shape[2:] for x in input):
+        raise ValueError("rand_sample() 'data' tensors must have identical spatial shape")
+    numel = shape[2:].numel()
+    if not replacement and num_samples > numel:
+        raise ValueError("rand_sample() 'num_samples' is greater than number of spatial points")
+    input = [x.flatten(2) for x in input]
+    if mask is None:
```
Comment (qiuhuaqi) [inline deepali/core/image.py:1210]: Right now the indices for different tensors in a batch would be the same when sampling without a mask and without replacement, while different for each tensor in other settings. Would it make sense to use `multinomial()` with a mask of equal weights for the no-mask setting? 
This shouldn't make much difference function-wise, but it would make the behavior more consistent and the code cleaner.

Comment (aschuh-hf) [reply deepali/core/image.py:1210]: Simplifying the code by always using `multinomial` with a default mask of ones makes sense. It would probably be best to time this. Note that the current implementation does indeed sample different indices for different samples in a batch for all code paths. Notice the `for row in index` loop of this modified code which produces a different permutation for each sample in the batch.

Comment (aschuh-hf) [reply deepali/core/image.py:1210]: I've done some simple timing measurements, and seems because of the loop `randperm` is considerably slower than just using `multinomial` with a mask of ones:

```
t_elapsed_1 = 0
for i in range(5):
    start.record()
    indices = U.rand_sample(data, num_samples, mask=None, replacement=False, generator=generator)
    end.record()
    torch.cuda.synchronize()
    if i > 0:
        t_elapsed_1 += start.elapsed_time(end)
t_elapsed_1 /= 4

t_elapsed_2 = 0
for i in range(5):
    start.record()
    mask = torch.ones((1, 1) + data.shape[2:], device=device)
    indices = U.rand_sample(data, num_samples, mask=mask, replacement=False, generator=generator)
    end.record()
    torch.cuda.synchronize()
    if i > 0:
        t_elapsed_2 += start.elapsed_time(end)
t_elapsed_2 /= 4
```

Gives
- t_elapsed_1 = 1.44 ms
- t_elapsed_2 = 0.35 ms


In case of `replacement=True`, calling `randint` is the simplest and fastest, though. So probably best to keep that `if` branch.

Comment (aschuh-hf) [reply deepali/core/image.py:1210]: For an input of larger size, hitting the limit of `torch.multinomial`, things look differently:

Input size `(5, 1, 256, 512, 512)`:
- t_elapsed_time_1 = 110 ms
- t_elapsed_time_2 = 6277 ms

The inverse transform sampling is considerably more expensive than a few `randperm` calls.

Comment (aschuh-hf) [reply deepali/core/image.py:1210]: Could also consider using https://github.com/LeviViana/torch_sampling which may be more performant.

Comment (qiuhuaqi) [reply deepali/core/image.py:1210]: Yes sorry I misread the randperm-copy loop. Indeed it's generating different index for different samples. I agree with the conclusion with what's shown from timing.
